# Lab: Building a Reproducible Analysis Workflow

In the first part of this lab, you will use official National Park Service data to answer a guided question. In the second part, you will find a dataset and conduct an analysis of your own.

Use the workflow from lecture throughout the lab:

1. **Question**
2. **Data**
3. **Operation**
4. **Check**
5. **Evidence**
6. **Conclusion**
7. **Limitation**


## Learning goals

By the end of this lab, you should be able to:

- organize an investigation with the reproducible analysis workflow;
- inspect a dataset before analyzing it;
- group, summarize, check, and sort observations with Pandas;
- formulate a research question that can be answered with data; and
- distinguish evidence, conclusions, and limitations.


## Using AI tools responsibly

You may use AI to ask questions, debug code, or receive feedback. You remain responsible for understanding your work, verifying suggestions against the data, and citing the source of your dataset. At the end, disclose whether you used an AI tool. Using AI is not required.


# Part 1: Florida park visitation


## Background

The National Park Service (NPS) publishes estimates of recreational visits to its units. These estimates help parks understand patterns in public use and plan staffing, maintenance, transportation, and visitor services.

This course file contains monthly estimates for five Florida NPS units from 2015 through 2025:

- Big Cypress National Preserve
- Biscayne National Park
- Canaveral National Seashore
- Dry Tortugas National Park
- Everglades National Park

A recreational visit is a visit, not necessarily one unique person. A person who visits several times may be counted several times.


| Feature | Description |
| --- | --- |
| `park_code` | Four-letter NPS unit code |
| `park` | Name of the NPS unit |
| `year` | Calendar year of the observation |
| `month_number` | Month represented as an integer from 1 through 12 |
| `month` | Month name |
| `recreation_visits` | Estimated recreational visits during that park-month |


## Step 1: Question

**After the decline in July 2020, when did Canaveral National Seashore's July recreational visitation first return to or exceed its average July visitation from 2015–2019?**

The 2015–2019 mean provides a pre-pandemic comparison point. In this question, “recovery” means the first July from 2020 onward with visitation at or above that baseline. This definition does not require every later July to remain above the baseline.


## Step 2: Data

Import Pandas using the alias `pd`. Read `data/nps_florida_park_visitation_2015_2025.csv` into a DataFrame named `park_visits`, then display its first five rows.


In [17]:
# Write your code here.
import pandas as pd

park_visits = pd.read_csv("data/nps_florida_park_visitation_2015_2025.csv")

park_visits.head()

,park_code,park,year,month_number,month,recreation_visits
0,BICY,Big Cypress National Preserve,2015,1,January,114045
1,BICY,Big Cypress National Preserve,2015,2,February,144112
2,BICY,Big Cypress National Preserve,2015,3,March,153084
3,BICY,Big Cypress National Preserve,2015,4,April,113880
4,BICY,Big Cypress National Preserve,2015,5,May,72510


Display the last five rows.


In [18]:
# Write your code here.
park_visits.tail()

,park_code,park,year,month_number,month,recreation_visits
655,EVER,Everglades National Park,2025,8,August,50464
656,EVER,Everglades National Park,2025,9,September,32290
657,EVER,Everglades National Park,2025,10,October,19008
658,EVER,Everglades National Park,2025,11,November,53416
659,EVER,Everglades National Park,2025,12,December,70926


Display information about the DataFrame and calculate the number of missing observations in each feature.


In [19]:
# Write your code here.
park_visits.info()

park_visits.isnull().sum()

<class 'pandas.DataFrame'>
RangeIndex: 660 entries, 0 to 659
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype
---  ------             --------------  -----
 0   park_code          660 non-null    str  
 1   park               660 non-null    str  
 2   year               660 non-null    int64
 3   month_number       660 non-null    int64
 4   month              660 non-null    str  
 5   recreation_visits  660 non-null    int64
dtypes: int64(3), str(3)
memory usage: 54.1 KB


park_code            0
park                 0
year                 0
month_number         0
month                0
recreation_visits    0
dtype: int64

**What does each row represent?**

Your answer: Each row represents the estimated recreational visits for one park during one specific month of one specific year (a single park-month observation.


**Are any observations missing from the six features?**

Your answer: No. The ".info()" output shows 660 non-null values in all six features, and ".isnull().sum()" returns 0 for for each column, so there are no missing observations.


## Step 3: Operation

Create one Boolean condition that identifies Canaveral National Seashore and another that identifies July.


In [20]:
# Write your code here.
canaveral_condition = park_visits["park"] == "Canaveral National Seashore"

july_condition = park_visits["month"] == "July"

Combine the two conditions with `&`. Use `.loc` to select the `year` and `recreation_visits` features, make a copy named `canaveral_july`, and display it.


In [21]:
# Write your code here.
canaveral_july = park_visits.loc[canaveral_condition & july_condition, ["year", "recreation_visits"]].copy()

canaveral_july

,year,recreation_visits
270,2015,163169
282,2016,216437
294,2017,138018
306,2018,189285
318,2019,162947
330,2020,112898
342,2021,206636
354,2022,233352
366,2023,173296
378,2024,176242


Create `pre_pandemic_condition`, which is `True` for years from 2015 through 2019. Use it with `.loc` to calculate the mean of `recreation_visits`, and store the result as `pre_pandemic_july_mean`.

Display `pre_pandemic_july_mean`.


In [22]:
# Write your code here.
pre_pandemic_condition = (canaveral_july["year"] >= 2015) & (canaveral_july["year"] <= 2019)

pre_pandemic_july_mean = canaveral_july.loc[pre_pandemic_condition, "recreation_visits"].mean()

pre_pandemic_july_mean

np.float64(173971.2)

Create a condition that identifies rows from 2020 onward whose `recreation_visits` are greater than or equal to `pre_pandemic_july_mean`. Use `.loc` to create `recovery_years`, then display it in ascending year order.


In [23]:
# Write your code here.
recovery_condition = (canaveral_july["year"] >= 2020) & (canaveral_july["recreation_visits"] >= pre_pandemic_july_mean)

recovery_years = canaveral_july.loc[recovery_condition, ["year", "recreation_visits"]].sort_values("year")

recovery_years

,year,recreation_visits
342,2021,206636
354,2022,233352
378,2024,176242
390,2025,253403


## Step 4: Check

Display the number of rows in `canaveral_july`, its smallest year, and its greatest year.


In [24]:
# Write your code here.
print("Number of rows:", len(canaveral_july))
print("Smallest year:", canaveral_july["year"].min())
print("Greatest year:", canaveral_july["year"].max())

Number of rows: 11
Smallest year: 2015
Greatest year: 2025


**Do these results confirm that the filtered data contains one July observation for every year from 2015 through 2025?**

Your answer: Yes. The check shows exactly 11 rows with a minimum year of 2015 and a maximum year of 2025. That means there is one July observation per year with no missing or duplicate years.


Count the observations used to calculate `pre_pandemic_july_mean`.


In [25]:
# Write your code here.
pre_pandemic_condition.sum()

np.int64(5)

**Does the baseline use the expected five years?**

Your answer: Yes. The "pre_pandemic_condition.sum()" returns 5, confirming the baseline averaged exactly the five years from 2015 through 2019.


## Step 5: Evidence

Display the first row of `recovery_years`.


In [26]:
# Write your code here.
recovery_years.head(1)

,year,recreation_visits
342,2021,206636


**What was the 2015–2019 mean, and which year first met or exceeded it after the July 2020 decline? Report that year's July visitation.**

Your answer: The 2015-2019 July mean was about 173,971 visits. After the July 2020 decline, 2021 was the first July to meet or exceed that baseline, with 206,636 recreational visits.


## Step 6: Conclusion


**Write one sentence that answers the research question using the evidence.**

Your answer: Canaveral National Seashore's July recreational visitation first returned to its 2015-2019 average in 2021, when it recorded 206,636 visits, exceeding the pre-pandemic baseline of about 173,971.


## Step 7: Limitation


**Does first exceeding the baseline mean that every later July remained above it? Use the data to explain.**

Your answer: No. Although 2021 was the first to meet the baseline, 2023 fell back below it, so it does not appear in "recovery_years". This shows that first exceeding baseline does not mean every later July stayed above it.


**Why does this analysis alone not establish that the pandemic caused the 2020 decline?**

Your answer: This analysis only shows that visitation dropped in 2020; it does not rule out other possible causes such as park closures, staffing changes, weather events, or differences in how visits were counted. Showing that the decline happened at the same time as the pandemic is correlation, not proof of causation, so this analysis alone cannot establish that the pandemiccaused the decline.


# Part 2: Your own investigation


## Find a dataset

Find a CSV dataset that interests you. Kaggle is one possible source, but you may use another reputable source. Choose a dataset that:

- has a clearly identified publisher or creator;
- includes documentation explaining the observations and features;
- contains at least one categorical and one quantitative feature;
- is appropriate to share in a class assignment;
- does not contain private or personally identifying information; and
- is small enough to run comfortably in this notebook.

Save the CSV in the notebook's `data` folder before continuing.


**What is the dataset's title, publisher or creator, and direct source URL?**

Your answer:
Title: Video games sales
Creator: elis3e
URL: https://www.kaggle.com/datasets/elis3e/video-games-sales


**What does one row represent?**

Your answer: Each row represents a single video game title released on a specific platform, along with its genre, publisher, release year, and sales figures by region.


## Step 1: Question

Formulate one research question that can be answered with your dataset. The question must require a calculation or comparison, not merely looking up one value.


**What is your research question?**

Your answer: Which video game genre has the highest average global sales?


## Step 2: Data

Read your CSV into a clearly named DataFrame. Display the first and last five rows, information about the DataFrame, and the missing-value count for every feature.


In [27]:
# Write your code here.
game_sales = pd.read_csv("data/vgsales.csv")

print("First five rows:")
display(game_sales.head())

print("Last five rows:")
display(game_sales.tail())

game_sales.info()

print("Missing values per column:")
game_sales.isnull().sum()

First five rows:


,Rank,Name,Platform,Year,Genre,Publisher,NA_Sales,EU_Sales,JP_Sales,Other_Sales,Global_Sales
0,259,Asteroids,2600,1980,Shooter,Atari,4.00,0.26,0.0,0.05,4.31
1,545,Missile Command,2600,1980,Shooter,Atari,2.56,0.17,0.0,0.03,2.76
2,1768,Kaboom!,2600,1980,Misc,Activision,1.07,0.07,0.0,0.01,1.15
3,1971,Defender,2600,1980,Misc,Atari,0.99,0.05,0.0,0.01,1.05
4,2671,Boxing,2600,1980,Fighting,Activision,0.72,0.04,0.0,0.01,0.77


Last five rows:


,Rank,Name,Platform,Year,Genre,Publisher,NA_Sales,EU_Sales,JP_Sales,Other_Sales,Global_Sales
16319,16565,Mighty No. 9,XOne,2016,Platform,Deep Silver,0.01,0.00,0.00,0.0,0.01
16320,16572,Resident Evil 4 HD,XOne,2016,Shooter,Capcom,0.01,0.00,0.00,0.0,0.01
16321,16573,Farming 2017 - The Simulation,PS4,2016,Simulation,UIG Entertainment,0.00,0.01,0.00,0.0,0.01
16322,16579,Rugby Challenge 3,XOne,2016,Sports,Alternative Software,0.00,0.01,0.00,0.0,0.01
16323,16592,Chou Ezaru wa Akai Hana: Koi wa Tsuki ni Shiru...,PSV,2016,Action,dramatic create,0.00,0.00,0.01,0.0,0.01


<class 'pandas.DataFrame'>
RangeIndex: 16324 entries, 0 to 16323
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Rank          16324 non-null  int64  
 1   Name          16324 non-null  str    
 2   Platform      16324 non-null  str    
 3   Year          16324 non-null  int64  
 4   Genre         16324 non-null  str    
 5   Publisher     16288 non-null  str    
 6   NA_Sales      16324 non-null  float64
 7   EU_Sales      16324 non-null  float64
 8   JP_Sales      16324 non-null  float64
 9   Other_Sales   16324 non-null  float64
 10  Global_Sales  16324 non-null  float64
dtypes: float64(5), int64(2), str(4)
memory usage: 2.1 MB
Missing values per column:


Rank             0
Name             0
Platform         0
Year             0
Genre            0
Publisher       36
NA_Sales         0
EU_Sales         0
JP_Sales         0
Other_Sales      0
Global_Sales     0
dtype: int64

**Which observations and features will you use to answer your question?**

Your answer: I will use the genre feature to group the games and the global_sales feature to calculate each genre's average sales. I will use all 16,324 rows, since neither genre nor global_sales has missing values.


## Step 3: Operation

Write the code needed to prepare, filter, group, summarize, or sort the data. Store the final evidence in a clearly named object and display it.


In [28]:
# Write your code here.
genre_avg_sales = game_sales.groupby("Genre")["Global_Sales"].mean().sort_values(ascending=False)

genre_avg_sales

Genre
Platform        0.946518
Shooter         0.800468
Role-Playing    0.628863
Racing          0.592798
Sports          0.568247
Fighting        0.531160
Action          0.529788
Misc            0.466444
Simulation      0.458472
Puzzle          0.424203
Strategy        0.258465
Adventure       0.184013
Name: Global_Sales, dtype: float64

**Explain how your operations connect the selected data to your research question.**

Your answer: My question asks which genre has the highest average global sales. Grouping by genre and taking the mean of global_sales calculates the average sales for each genre, and sorting in descending order puts the highest selling genre first, which directly answers the question.


## Step 4: Check

Write at least one check that could reveal an error in your analysis, then display its result.


In [29]:
# Write your code here.
games_per_genre = game_sales.groupby("Genre").size()
print("Games per genre:")
print(games_per_genre)
print("Total games across all genres:", games_per_genre.sum())
print("Total rows in dataset:", len(game_sales))

Games per genre:
Genre
Action          3252
Adventure       1276
Fighting         836
Misc            1710
Platform         876
Puzzle           571
Racing          1226
Role-Playing    1469
Shooter         1282
Simulation       851
Sports          2304
Strategy         671
dtype: int64
Total games across all genres: 16324
Total rows in dataset: 16324


**What did you check, and did the result meet your expectation?**

Your answer: I checked that the total number of games across all genres equals the total rows in the dataset. Both came to 16,324, which met my expectation and confirms the group-by used every game exactly once, with none dropped or duplicated.


## Step 5: Evidence


**Which specific value or comparison in your output answers the question?**

Your answer: The average global sales per genre in "genre_avg_sales" answers the question. Platform has the highest average at about 0.95 million sales per game, which is higher than every other genre.


## Step 6: Conclusion


**What conclusion is supported by the evidence?**

Your answer: Based on the dataset, the platform genre has the highest average global sales per game, making it the top selling genre on average.


## Step 7: Limitation


**What should a reader avoid concluding from your analysis, and why?**

Your answer: A reader should not conclude that platform is the most popular or the highest earning genre overall. This is an average per game, and platform has relatively few games, so a high average doesn't mean high total sales. A genre with many games could earn more in total. The analysis also doesn't account for release year, region, or the age of the dataset, so it shouldn't be read as reflecting current game sales.


## AI-use reflection


**State whether you used an AI tool. If you did, describe one suggestion you checked and what you accepted, changed, or rejected. If you did not, state that you did not use one.**

Your response: Yes, I used an AI tool. One suggestion was to load the CSV with "pd.read_csv("data/vgsales.csv")", which gave me a file not found error. I checked it by running "os.listdir("data")", and found the file wasn't in the folder my the notebook was using. I then moved the file to the correct folder and the code worked. I accepted the approach but had to fix the file location myself.


## Submission checklist

Before submitting, confirm that all cells run in order, the independent dataset and its source are included, every answer is complete, numerical claims match the output, conclusions are appropriately limited, and the AI-use reflection is complete.
